pubsub_logging/async_handler.py

```diff
@@ -0,0 +1,145 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Python logging handler implementation for Cloud Pub/Sub.
+
+This module provides a logging.Handler implementation which sends the
+logs to Cloud Pub/Sub[1] asynchronously. The logs are kept in an
+internal queue and child workers will pick them up and send them in
+background.
+
+[1]: https://cloud.google.com/pubsub/docs
+
+"""
+
+import logging
+import multiprocessing as mp
+
+# For Python 2 and Python 3 compatibility.
+try:
+    from queue import Empty
+except ImportError:
+    from Queue import Empty
+
+from pubsub_logging import errors
+
+from pubsub_logging.utils import compat_urlsafe_b64encode
+from pubsub_logging.utils import get_pubsub_client
+from pubsub_logging.utils import publish_body
+
+
+BATCH_SIZE = 1000
+DEFAULT_WORKER_SIZE = 1
+DEFAULT_RETRY_COUNT = 10
+DEFAULT_TIMEOUT = 1
+
+
+class AsyncPubsubHandler(logging.Handler):
+    """A logging handler to publish logs to Cloud Pub/Sub in background."""
+    def __init__(self, topic, worker_size=DEFAULT_WORKER_SIZE,
+                 retry=DEFAULT_RETRY_COUNT, timeout=DEFAULT_TIMEOUT,
+                 client=None, publish_body=publish_body, stderr_logger=None):
+        """The constructor of the handler.
+
+        Args:
+          topic: Cloud Pub/Sub topic name to send the logs.
+          worker_size: The initial worker size, defaults to 1.
+          retry: How many times to retry upon Cloud Pub/Sub API failure,
+                 defaults to 5.
+          timeout: Timeout for BatchQueue.get, defaults to 1.
+          client: An optional Cloud Pub/Sub client to use. If not set, one is
+                  built automatically, defaults to None.
+          publish_body: A callable for publishing the Pub/Sub message,
+                        just for testing purposes.
+          stderr_logger: A logger for informing failures with this
+                         logger, defaults to None and if not specified, a last
+                         resort logger will be used.
+        """
+        super(AsyncPubsubHandler, self).__init__()
+        self._topic = topic
+        self._retry = retry
+        self._client = client
+        self._q = mp.JoinableQueue()
+        self._should_exit = mp.Value('i', 0)
+        self._workers = []
+        self._timeout = timeout
+        self._threshold = BATCH_SIZE
+        self._buf = []
+        self._publish_body = publish_body
+        if stderr_logger:
+            self._stderr_logger = stderr_logger
+        else:
+            self._stderr_logger = logging.Logger('last_resort')
+            self._stderr_logger.addHandler(logging.StreamHandler())
+        for _ in range(worker_size):
+            p = mp.Process(target=self.send_loop, args=(self._q,))
+            p.daemon = True
+            self._workers.append(p)
+            p.start()
+
+    def send_loop(self, q):  # pragma: NO COVER
```

Comment:
I'm of the opinion that this probably doesn't need to be an instance method, since it really only needs to be given a queue and a client. I'd probably make it a static method on AsyncPububHandler or just move it into the top level of this file. That way your code becomes:

```py
self._pool = mp.Pool(pool_size, self.send_loop, (self._client, self._q))
```

```py
def ExportToPubsub(client, q):
```